### DISCUSSION TITLE
Modeling assignment with unique task sets

### DISCUSSION THREAD

**Discussion by rroblak** [google/or-tools#3667] [CP-SAT questions]
Hi,

Thanks for the amazing library. I'm trying to model an assignment problem with the following characteristics:

- Constraint 1: Each worker must be assigned at least 1 and not more than _xi_ tasks, where _xi_ is pre-determined limit for worker _i_.
- Constraint 2: Each worker's task set must not equal any other worker's task set. They can overlap, subset, etc. but they must not be equal.
- Objective: Minimize the average absolute deviation of worker count per task (i.e., ideally each task should have the same number of workers assigned to it).
- Scale:
  - Workers: ~100k
  - Tasks: ~1k
  - Maximal number of tasks for a worker: ~333

I was able to model _Constraint 1_ and the _Objective_ with the CP-SAT solver.

However, I'm struggling to model _Constraint 2_.  I was hoping to map task sets to integers and use `#addAllDifferent()` (Java), but I believe that will maximally support task sets of size 64 (due to `int64`; technique is to convert the boolean array representing a task set to an integer).

Should I be using a different solver to determine the optimal edges? Or is there a different technique I can use?

Thanks for any help! 🙏

**Comment by Regista6** [comment]
Mapping task sets to integers and then using AddAllDifferent seems like the way to go (imo).  "maximally support task sets of size 64" can you elaborate on this?.
